server/src/main/scala/com/stratio/crossdata/server/CrossdataServer.scala

```diff
@@ -16,38 +16,229 @@
 import akka.http.scaladsl.Http.ServerBinding
 import akka.routing.{DefaultResizer, RoundRobinPool}
-import akka.stream.ActorMaterializer
+import com.hazelcast.config.{XmlConfigBuilder, Config => HzConfig}
 import com.stratio.crossdata.common.util.akka.keepalive.KeepAliveMaster
 import com.stratio.crossdata.server.actors.{ResourceManagerActor, ServerActor}
 import com.stratio.crossdata.server.config.ServerConfig
-import org.apache.commons.daemon.{Daemon, DaemonContext}
+import com.stratio.crossdata.server.discovery.{ServiceDiscoveryConfigHelper => SDCH, ServiceDiscoveryHelper => SDH}
+import com.typesafe.config.{Config, ConfigValueFactory}
+import org.apache.curator.framework.recipes.leader.LeaderLatch
+import org.apache.curator.framework.{CuratorFramework, CuratorFrameworkFactory}
+import org.apache.curator.retry.ExponentialBackoffRetry
+import org.apache.curator.utils.ZKPaths
 import org.apache.log4j.Logger
 import org.apache.spark.sql.crossdata
 import org.apache.spark.sql.crossdata.session.{BasicSessionProvider, HazelcastSessionProvider, XDSessionProvider}
 import org.apache.spark.{SparkConf, SparkContext}
 
 import scala.collection.JavaConversions._
-import scala.concurrent.Future
+import scala.concurrent.{ExecutionContext, Future}
+import scala.concurrent.duration.FiniteDuration
+import scala.util.{Random, Try}
 
 
-class CrossdataServer extends ServerConfig {
+class CrossdataServer(progrConfig: Option[Config] = None) extends ServerConfig {
 
   override lazy val logger = Logger.getLogger(classOf[CrossdataServer])
 
   var system: Option[ActorSystem] = None
   var sessionProviderOpt: Option[XDSessionProvider] = None
   var bindingFuture: Option[Future[ServerBinding]] = None
 
+  private val serverConfig = progrConfig map (_.withFallback(config)) getOrElse (config)
+
+  private val hzConfig: HzConfig = new XmlConfigBuilder().build()
+
+  private def startDiscoveryClient(sdConfig: SDCH): CuratorFramework = {
+
+    val curatorClient = CuratorFrameworkFactory.newClient(
+      sdConfig.getOrElse(SDCH.ServiceDiscoveryUrl, SDCH.ServiceDiscoveryDefaultUrl),
+      new ExponentialBackoffRetry(
+        SDCH.RetrySleep,
+        SDCH.Retries))
+    curatorClient.start
+    curatorClient.blockUntilConnected
+    curatorClient
+  }
+
+  private def requestSubscriptionLeadership(dClient: CuratorFramework, sdc: SDCH) = {
+
+    val sLeaderPath = sdc.getOrElse(SDCH.SubscriptionPath, SDCH.DefaultSubscriptionPath)
+
+    logger.debug(s"Service discovery - subscription leadership path: $sLeaderPath")
+
+    ZKPaths.mkdirs(dClient.getZookeeperClient.getZooKeeper, sLeaderPath)
+
+    val sLeader = new LeaderLatch(dClient, sLeaderPath)
+
+    sLeader.start
+
+    Try(sLeader.await(
+      sdc.getOrElse(SDCH.SubscriptionTimeoutPath, SDCH.DefaultSubscriptionTimeout.toString).toLong, TimeUnit.SECONDS))
+      .getOrElse(throw new RuntimeException(
+      "Crossdata Server cannot be started because access to service discovery is blocked"))
```

Comment:
Following the comment above, this `Try` should be the return value and could change the network or timeout exception to ` new RuntimeException("Crossdata Server cannot be started because access to service discovery is blocked")` using `recoverWith`